substrate/frame/revive/src/limits.rs

```diff
@@ -58,3 +61,82 @@ pub const STORAGE_KEY_BYTES: u32 = 128;
 ///
 /// The buffer will always be disabled for on-chain execution.
 pub const DEBUG_BUFFER_BYTES: u32 = 2 * 1024 * 1024;
+
+/// The page size in which PolkaVM should allocate memory chunks.
+pub const PAGE_SIZE: u32 = 4 * 1024;
+
+/// Limits that are only enforced on code upload.
+///
+/// # Note
+///
+/// This limit can be increased later without breaking existing contracts
+/// as it is only enforced at code upload time. Code already uploaded
+/// will not be affected by those limits.
+pub mod code {
+	use super::PAGE_SIZE;
+	use crate::{Config, Error, LOG_TARGET};
+	use frame_support::ensure;
+	use sp_runtime::DispatchResult;
+
+	/// The maximum length of a code blob in bytes.
+	///
+	/// This mostly exist to prevent parsing too big blobs and to
+	/// have a maximum encoded length. The actual memory calculation
+	/// is purely based off [`STATIC_MEMORY_BYTES`].
+	pub const BLOB_BYTES: u32 = 256 * 1024;
+
+	/// Maximum size the program is allowed to take in memory.
+	///
+	/// This includes data and code. Increasing this limit will allow
+	/// for more code or more data. However, since code will decompress
+	/// into a bigger representation on compilation it will only increase
+	/// the allowed code size by [`BYTE_PER_INSTRUCTION`].
+	pub const STATIC_MEMORY_BYTES: u32 = 1024 * 1024;
+
+	/// How much memory each instruction will take in-memory after compilation.
+	///
+	/// This is `size_of<usize>() + 16`. But we don't use `usize` here so it isn't
+	/// different on the native runtime (used for testing).
+	const BYTE_PER_INSTRUCTION: u32 = 20;
+
+	/// The code is stored multiple times as part of the compiled program.
+	const EXTRA_OVERHEAD_PER_CODE_BYTE: u32 = 4;
+
+	/// Make sure that the various program parts are within the defined limits.
+	pub fn enforce<T: Config>(blob: &[u8]) -> DispatchResult {
+		fn round_page(n: u32) -> u32 {
+			debug_assert!(
+				PAGE_SIZE != 0 && (PAGE_SIZE & (PAGE_SIZE - 1)) == 0,
+				"Page size must be power of 2"
+			);
+			(n + PAGE_SIZE - 1) & !(PAGE_SIZE - 1)
+		}
+
+		let program = polkavm::ProgramBlob::parse(blob.into()).map_err(|err| {
+			log::debug!(target: LOG_TARGET, "failed to parse polkavm blob: {err:?}");
+			Error::<T>::CodeRejected
+		})?;
+
+		// this is O(n) but it allows us to be more precise
+		let num_instructions = program.instructions().count();
+
+		// The memory consumptions is the byte size of the whole blob,
+		// minus the RO data payload in the blob,
+		// minus the RW data payload in the blob,
+		// plus the RO data in memory (which is always equal or bigger than the RO payload),
+		// plus RW data in memory, plus stack size in memory.
+		// plus the overhead of instructions in memory which is derived from the code
+		// size itself and the number of instruction
+		let memory_size = blob.len() as u64 + round_page(program.ro_data_size()) as u64 -
+			program.ro_data().len() as u64 +
+			round_page(program.rw_data_size()) as u64 -
```

Comment:
Probably the safe bet to use checked math here; I think if we eventually switch to 64bit, the upcast to u64 here no longer prevents malicious inputs from panicing it

Reply:
Yes assuming that `ro_data_size` and `ro_data_len` will be changed to `u64` this code could become insecure later on without any compilation failure.  I think we should play it safe and just do checked math in `u32`. Our limit is in `u32` and will stay there.

Reply:
Using checked math in `u64` now. Probably not necessary in all places but I think this makes it easier to argue that the code is correct.

Reply:
Yes I agree it looks easy to reason about and it is a one time costs so should not matter :)